cdap-master/src/main/java/co/cask/cdap/master/environment/k8s/AbstractServiceMain.java

```diff
@@ -0,0 +1,262 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.master.environment.k8s;
+
+import co.cask.cdap.api.metrics.MetricsCollectionService;
+import co.cask.cdap.common.app.MainClassLoader;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.guice.ConfigModule;
+import co.cask.cdap.common.guice.DFSLocationModule;
+import co.cask.cdap.common.guice.IOModule;
+import co.cask.cdap.common.logging.LoggingContext;
+import co.cask.cdap.common.logging.LoggingContextAccessor;
+import co.cask.cdap.common.options.Option;
+import co.cask.cdap.common.options.OptionsParser;
+import co.cask.cdap.common.runtime.DaemonMain;
+import co.cask.cdap.common.utils.ProjectInfo;
+import co.cask.cdap.logging.appender.LogAppender;
+import co.cask.cdap.logging.appender.LogAppenderInitializer;
+import co.cask.cdap.logging.appender.LogMessage;
+import co.cask.cdap.master.environment.DefaultMasterEnvironmentContext;
+import co.cask.cdap.master.environment.MasterEnvironmentExtensionLoader;
+import co.cask.cdap.master.spi.environment.MasterEnvironment;
+import co.cask.cdap.metrics.guice.MetricsClientRuntimeModule;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+import com.google.common.util.concurrent.Service;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import com.google.inject.Module;
+import com.google.inject.Provider;
+import com.google.inject.Scopes;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.discovery.DiscoveryService;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * The abstract base class for writing various CDAP master service main classes.
+ */
+public abstract class AbstractServiceMain extends DaemonMain {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractServiceMain.class);
+
+  private final List<Service> services = new ArrayList<>();
+  private final List<AutoCloseable> closeableResources = new ArrayList<>();
+  private Injector injector;
+  private MasterEnvironment masterEnv;
+
+  /**
+   * Helper method for sub-class to call from static void main.
+   *
+   * @param mainClass the class of the master main class implementation
+   * @param args arguments to main
+   * @param <T> type of the master main class
+   * @throws Exception if execution failed
+   */
+  protected static <T extends AbstractServiceMain> void main(Class<T> mainClass, String[] args) throws Exception {
+    ClassLoader classLoader = MainClassLoader.createFromContext();
```

Comment:
why is this needed?

Reply:
This is for class interception and rewriting purpose. It is used for rewriting Dataset classes as well as classes annotated with the `AuthEnforce` annotation.

All CDAP processes have to be using the MainClassLoader as the system classloader. We do the same in StandaloneMain as well as for all master TwillRunnables, although for the twill case, it is handled by twill and we only need to provide the class name of the ClassLoader to use.